DataCleansing example application

anwar6953 opened the pull request:
Requesting review of the idea and overall implementation.

Main idea is to demonstrate incremental processing of the partitions of a PartitionedFileSet, using the concept of data validation to filter all records that do not conform to a given schema. Currently, the schema is defined within the mapper, but an idea is to change this to accept this via runtime arguments.

Suggestions in regards to the data schema or validation are welcome.

https://issues.cask.co/browse/CDAP-2994
http://builds.cask.co/browse/CDAP-RBT358-6

jayfjin commented on cdap-examples/DataCleansing/pom.xml:
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  Copyright © 2014 Cask Data, Inc.
```
copyright year

jayfjin commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansing.java:
```diff
@@ -0,0 +1,63 @@
+    createDataset(INPUT_RECORDS, PartitionedFileSet.class, PartitionedFileSetProperties.builder()
+      // Properties for partitioning
+      .setPartitioning(Partitioning.builder().addLongField("time").build())
+        // Properties for file set
```
indentation? also line 59

jayfjin commented on cdap-examples/DataCleansing/src/test/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduceTest.java:
```diff
@@ -0,0 +1,145 @@
+    return records;
+  }
+
+  //
```
remove

jayfjin commented on cdap-examples/DataCleansing/src/test/java/co/cask/cdap/examples/datacleansing/SimpleSchemaMatcherTest.java:
```diff
@@ -0,0 +1,89 @@
+import org.junit.Test;
+
+public class SimpleSchemaMatcherTest {
+    Schema simplePersonSchema = Schema.recordOf("person",
```
can we use DataCleansingMapReduce.SchemaMatchingFilter.SCHEMA and remove this?

anwar6953 replied:
I'm considering removing the schema from the mapreduce, so that user can pass their own via runtime arguments, so I will leave this here until that's finalized.

jayfjin commented on cdap-examples/DataCleansing/src/test/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduceTest.java:
```diff
@@ -0,0 +1,145 @@
+    mapReduceManager.waitForFinish(5, TimeUnit.MINUTES);
+
+    Assert.assertEquals(Collections.<String>emptySet(), getCleanData(serviceURL, now));
+
```
remove newline

jayfjin commented on cdap-examples/DataCleansing/src/test/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduceTest.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  private Set<String> splitToRecords(String content) {
+    Set<String> records = new HashSet<>();
```
Sets.newHashSet()?

jayfjin commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduce.java:
```diff
@@ -0,0 +1,135 @@
+    // Each run writes its output to a partition for the league
+    Long timeKey = Long.valueOf(context.getRuntimeArguments().get(OUTPUT_PARTITION_TIME));
+    PartitionKey outputKey = PartitionKey.builder().addLongField("time", timeKey).build();
+    Map<String, String> outputArgs = Maps.newHashMap();
```
new HashMap();

anew commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansing.java:
```diff
@@ -0,0 +1,63 @@
+    createDataset("consumingState", KeyValueTable.class);
+
+    // Create the "inputRecords" partitioned file set for storing the input records, configure it to work with MapReduce
+    createDataset(INPUT_RECORDS, PartitionedFileSet.class, PartitionedFileSetProperties.builder()
```
why use constants for these dataset names, but not for consumingState?

rohitsinha54 commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansing.java:
```diff
@@ -0,0 +1,63 @@
+
+  @Override
+  public void configure() {
+    setName("DataCleansing");
```
Not needed if same as classname

anwar6953 replied:
It seems like a couple of other example apps do this, perhaps just to show the API being used

rohitsinha54 commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduce.java:
```diff
@@ -0,0 +1,135 @@
+
+  @Override
+  public void configure() {
+    setName(NAME);
```
Not needed and in other places you can do getSimpleName

anwar6953 replied:
I think it's better not to expose that internal detail to users, and do this instead?

rohitsinha54 replied:
The name shows on the UI and this is a public class so what internal detail are we trying to hide ?

anwar6953 replied:
the detail that a program's name defaults to its class's simple name.

anew replied:
Let's make the examples look nice and set the name. It is a convenience for users who are "lazy" that we fall back to the class name. But that is not necessarily pretty...

rohitsinha54 replied:
How about setName(DataCleansingMapReduce.class.getSimpleName())
- Sets the name explicitly
- Hides the details that it will fall back to class name. 
:)

anwar6953 replied:
I think its ok to stick with using constant, since it seems cleaner than getting the class's simplename. Also, many other example apps do it with that style. Either way, it's probably not a big deal.

rohitsinha54 commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduce.java:
```diff
@@ -0,0 +1,135 @@
+  @Override
+  public void configure() {
+    setName(NAME);
+    setMapperResources(new Resources(1024));
```
Why ?

anwar6953 replied:
We had done this in a couple of other example apps: WordCount, PurchaseApp.

anew replied:
Let's make the examples look nice and set the name. It is a convenience for users who are "lazy" that we fall back to the class name. But that is not necessarily pretty...

rohitsinha54 commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingService.java:
```diff
@@ -0,0 +1,108 @@
+
+  @Override
+  protected void configure() {
+    setName(NAME);
```
same

anwar6953 replied:
I don't think its a good idea to expose the internal detail that the default name is the class's simple name, within an example app

rohitsinha54 replied:
See above.

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+.. meta::
+    :author: Cask Data, Inc.
+    :description: Cask Data Application Platform DataCleansing Application
+    :copyright: Copyright © 2014 Cask Data, Inc.
```
2015

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+.. meta::
+    :author: Cask Data, Inc.
+    :description: Cask Data Application Platform DataCleansing Application
```
Data Cleansing (as this is a descriptive entry)

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+==============
+
+A Cask Data Application Platform (CDAP) example demonstrating incrementally consuming partitions of a partitioned file
+set using MapReduce.
```
file set -> fileset [common]

anew replied:
I am sure I agree. PartitionedFileSet either decomposes to partitioned file set or remains compound.

awholegunch replied:
Did you (andreas) mean you think we should use "file set", and not "fileset"? I though we preferred "file set".

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+Overview
+========
+
+This application has a MapReduce which processes records from a partitioned file set into another partitioned file set,
```
from one partitioned fileset

anwar6953 replied:
also recommend removing the space between 'file' and 'set' here, too?

anew replied:
that instead of which

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+As in the other :ref:`examples <examples-index>`, the components
+of the application are tied together by the class ``DataCleansing``:
+
+.. literalinclude:: /../../../cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansing.java
```
To guard against changes, add a has to examples/build.sh:
```
test_an_include 6b33a4fe12b6bed2742c4ce0bbe5864a ../../cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansing.java
```

anwar6953 replied:
can you describe how you created this hash?

awholegunch replied:
Sure. Add the line to the build.sh, with any hash in it. Run the docs build.sh command. You can either do it at the top level (cdap-docs/build.sh docs) or from within a manual's directory (examples-manual/build.sh docs).

In the output there will be an error message; it will have the correct hash. Copy and paste...

anwar6953 replied:
done

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+Building and Starting
+=====================
+
+.. include:: building-and-starting.txt
```
building-starting-running-cdap.txt

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+
+The MapReduce must be started with a runtime argument 'output.partition.key' in order to specify the output partition of
+the 'outputRecords' dataset to write to. In this example, we'll simply use 1 as the value.
+- Go to the *DataCleansing* `application overview page
```
Insert blank line before "- Go to the..."; Sphinx is confused.

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+Starting the MapReduce
+----------------------
+
+The MapReduce must be started with a runtime argument 'output.partition.key' in order to specify the output partition of
```
```
runtime argument ``output.partition.key`` in order
```

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+----------------------
+
+The MapReduce must be started with a runtime argument 'output.partition.key' in order to specify the output partition of
+the 'outputRecords' dataset to write to. In this example, we'll simply use 1 as the value.
```
```
the ``outputRecords`` dataset to write to. In this example, we'll simply use ``1`` as the value.
```

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+- Go to the *DataCleansing* `application overview page
+  <http://localhost:9999/ns/default/apps/DataCleansing/overview/status>`__,
+  click ``DataCleansingMapReduce`` to get to the MapReduce detail page, set the runtime
+  arguments 'output.partition.key' = 1, then click the *Start* button; or
```
```
arguments ``output.partition.key = 1``, then click the *Start* button; or
```
not certain about this, as it would depend exactly on how it looks in the UI.

anwar6953 replied:
I did this:
arguments using ``output.partition.key`` as the key and ``1`` as the value, then click

awholegunch commented on cdap-docs/examples-manual/source/examples/index.rst:
```diff
@@ -36,7 +36,10 @@ In addition to the :ref:`Getting Started's <getting-started-index>`
+  * - :doc:`Data Cleansing<data-cleansing>`
+    - TODO: A simple HelloWorld App that's written using CDAP. It introduces how the components stream, flow, dataset,
+      and service are used in a CDAP application.
+      * - :doc:`Hello World<hello-world>`
```
This is messed up. Line 42 needs to be flush left; the TODO needs to go, and the description needs rewriting...Data Cleansing should go after "Count Random"; HelloWorld should probably be th first example, and then all examples after that can be alphabetical...

awholegunch commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/SimpleSchemaMatcher.java:
```diff
@@ -0,0 +1,62 @@
+
+  /**
+   * Determines whether or not this matcher's schema fits a piece of data.
+   * A failure to match could arise from any of the following reasons:
```
A failure to match could arise from any of:

awholegunch commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/package-info.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * This DataCleansing Application consists of the following programs and datasets:
+ *
+ *   1. ``DataCleansingService`` that allows reading and writing to the ``inputRecords`` and ``outputRecords``
```
This HTML, not rst. Items with back-ticks should be (example) {@link DataCleansingService}

awholegunch replied:
And we could fix the other examples with this at the same time.

anwar6953 replied:
Done. Can you take a look and see if the update is appropriate?
Thanks.

awholegunch commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/package-info.java:
```diff
@@ -0,0 +1,35 @@
+ * the License.
+ */
+/**
+ * This package contains the DataCleansing sample Application filters records that do not match a given schema.
```
the DataCleansing Application that filters records that do not match a given schema.

awholegunch commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/package-info.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * This package contains the DataCleansing sample Application filters records that do not match a given schema.
+ *
+ * This DataCleansing Application consists of the following programs and datasets:
```
This DataCleansing Application consists of these programs and datasets:

awholegunch commented on cdap-examples/DataCleansing/src/test/java/co/cask/cdap/examples/datacleansing/SimpleSchemaMatcherTest.java:
```diff
@@ -0,0 +1,89 @@
+
+    Assert.assertTrue((schemaMatcher.matches(jsonObject.toString())));
+
+
```
remove duplicate empty lines, right?

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,170 @@
+
+    $ cdap-cli.sh call service DataCleansing.DataCleansingService GET /outputRecords/1
+
+The records that are not filtered out (adhere to the given schema) will be displayed::
```
``The records that are not filtered out (those that adhere to the given schema) will be displayed::``

Is this what you meant?

anwar6953 replied:
yea, replaced.

anew commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,178 @@
+
+- ``inputRecords`` input PartitionedFileSet of the DataCleansingMapReduce, contains any ingested records.
+- ``outputRecords`` output PartitionedFileSet, contains only the filtered records.
+- ``consumingState`` stores the state of the DataCleansingMapReduce, so each run, it processes only new partitions.
```
grammar? ...such that in each run, it processes...

anew commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,178 @@
+Data Storage
+------------
+
+- ``inputRecords`` input PartitionedFileSet of the DataCleansingMapReduce, contains any ingested records.
```
should we call this rawRecords and cleanRecords?

anwar6953 replied:
sure, will do.

anew commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,178 @@
+  arguments 'output.partition.key' = 1, then click the *Start* button; or
+- Use the Command Line Interface::
+
+    $ cdap-cli.sh start mapreduce DataCleansing.DataCleansingMapReduce output.partition.key=1
```
it would be nice if the caller did not have to specify the output key. Can we just make the output a TPFS?

anwar6953 replied:
If the output key is the current time, it's difficult for the user to have a handle on the exact partition (for retrieval purposes) from which a MR run outputted to.

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,178 @@
+specified as an additional runtime argument to the MapReduce with the key 'schema.key'.
+Otherwise, the default schema that is matched against the records:
+
+.. literalinclude:: /../../../cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduce.java
```
This literal include should be added to the build.sh with an updated hash, as per above instructions.

anwar6953 replied:
done

anew commented:
+1 for this example. I like the idea and it it illustrates a very common use case. Please see my comments.

bdmogal commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduce.java:
```diff
@@ -0,0 +1,156 @@
+    if (succeeded) {
+      partitionConsumer.persist(context);
+    }
+    super.onFinish(succeeded, context);
```
why do we need this?

anwar6953 replied:
If we did not override the `onFinish` method, the method in the super class would be called, so I am trying to maintain that behavior.

bdmogal replied:
But does it do anything in the super class? Otherwise its just a no-op method call?

anwar6953 replied:
I just want to maintain the behavior (of before I overrode the method) that if anything was put into the onFinish method, it would get executed.

anwar6953 replied:
removed.

bdmogal commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduce.java:
```diff
@@ -0,0 +1,156 @@
+        schemaMatcher = new SimpleSchemaMatcher(Schema.parseJson(schemaJson));
+      }
+
+      super.setup(context);
```
why do we need this?

anwar6953 replied:
removed.

bdmogal commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingService.java:
```diff
@@ -0,0 +1,106 @@
+  public static class RecordsHandler extends AbstractHttpServiceHandler {
+
+    @UseDataSet(DataCleansing.INPUT_RECORDS)
+    private PartitionedFileSet inputRecords;
```
Add ``@SuppressWarnings("unused") for both these?

bdmogal commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/SimpleSchemaMatcher.java:
```diff
@@ -0,0 +1,62 @@
+    // rely on validations in the StructuredRecord's builder
+    try {
+      JsonObject jsonObject = new JsonParser().parse(data).getAsJsonObject();
+      StructuredRecord.Builder builder = StructuredRecord.builder(schema);
```
Do you have to build the ``StructuredRecord`` every single time? Or would it be sufficient cache this object and only make sure that ``convertAndSet`` does not throw an exception for any key/value pair?

anwar6953 replied:
The fields added to the builder are checked against upon the `build` method call, so we can not use the same `StructuredRecord.Builder` object each time.

bdmogal replied:
:+1:

bdmogal commented on cdap-examples/DataCleansing/src/test/java/co/cask/cdap/examples/datacleansing/SimpleSchemaMatcherTest.java:
```diff
@@ -0,0 +1,89 @@
+    jsonObject.addProperty("dob", "5-20-92");
+
+    Assert.assertTrue((schemaMatcher.matches(jsonObject.toString())));
+
```
empty line

bdmogal commented on cdap-examples/DataCleansing/src/test/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduceTest.java:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  private void createPartition(URL serviceUrl, Set<String> records) throws IOException {
+    HttpResponse response = HttpRequests.execute(HttpRequest.post(new URL(serviceUrl, DataCleansing.INPUT_RECORDS))
```
nit: Perhaps split this to make more readable?

anwar6953 replied:
done.

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,179 @@
+---------------------------------
+
+``DataCleansingMapReduce`` is a simple MapReduce that reads from the ``rawRecords`` PartitionedFileSet and writes to
+the ``cleanRecords`` PartitionedFileSet. The ``beforeSubmit()`` method prepares the MapReduce program for this:
```
having the "()" is redundant with the word "method"...how about 
```
The ``beforeSubmit`` method prepares the MapReduce program:
```

anwar6953 replied:
agreed.

awholegunch commented on cdap-docs/examples-manual/source/examples/data-cleansing.rst:
```diff
@@ -0,0 +1,179 @@
+
+- It uses the ``BatchPartitionConsumer`` to specify the partitions to process as input, in order to only process new
+  partitions since its last run.
+- It specifies the output partition to write to based upon the runtime arguments.
```
It specifies the output partition that is written to, based upon the supplied runtime arguments.

awholegunch commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/package-info.java:
```diff
@@ -0,0 +1,42 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
```
Insert a blank line between these blocks

anwar6953 replied:
done

awholegunch commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/package-info.java:
```diff
@@ -0,0 +1,42 @@
+ */
+/**
+ * This package contains the DataCleansing Application that filters records that do not match a given schema.
+ *
```
```
add a <p> opening tag

awholegunch commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/package-info.java:
```diff
@@ -0,0 +1,42 @@
+ * This package contains the DataCleansing Application that filters records that do not match a given schema.
+ *
+ * This DataCleansing Application consists of these programs and datasets:
+ *
```
```
add </p>

anwar6953 replied:
Wait, does the <p> </p> enclose the first two lines of text?

awholegunch commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/package-info.java:
```diff
@@ -0,0 +1,42 @@
+ *   </li>
+ *
+ *   <li>
+ *   Three Datasets used by the MapReduce and Service:
```
```
This is a new list, but an <ul> list; it need opening, closing and <li>...</li> tags...

awholegunch commented:
There looks like there are some issues remaining regarding the naming methods, but aside from that, this LGTM.

anwar6953 commented:
What do you mean, issues regarding the naming methods, @awholegunch ?

bdmogal commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingService.java:
```diff
@@ -0,0 +1,108 @@
+  /**
+   * A handler that allows writing to the 'rawRecords' and reading from the 'cleanRecords' PartitionedFileSet.
+   */
+  public static class RecordsHandler extends AbstractHttpServiceHandler {
```
I think it would be good to make this handler more RESTful. Some suggestions:
1. Add a leading ``/v1`` in the Path
2. Perhaps make the APIs as ``/v1/records/raw`` and ``/v1/records/clean/{time}``

bdmogal commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/package-info.java:
```diff
@@ -0,0 +1,47 @@
+ *   </li>
+ *
+ *   <li>
+ *   A MapReduce named DataCleansingMapReduce that reads the files from a PartitionedFileSet, applies a filter
```
for consistency, link to ``DataCleansingMapReduce``?

anwar6953 commented:
@awholegunch Were you referring to this?
https://github.com/caskdata/cdap/pull/3231#discussion_r35606282

bdmogal commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingService.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @GET
+    @Path("records/clean/{time}")
```
Add a leading slash

bdmogal commented:
1 comment, then LGTM

chtyim commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingService.java:
```diff
@@ -0,0 +1,109 @@
+        responder.sendError(400, String.format("Unable to read path '%s'", partition.getRelativePath()));
+        return;
+      }
+      responder.send(200, content, "text/plain", ImmutableMultimap.<String, String>of());
```
Better move this into the try block and also move the ByteBuffer content to where it is used.

anwar6953 commented:
@chtyim @sreevatsanraman 
In the last commit, I removed the service's endpoint to serve data from the partitioned file set and removed the restriction that there is only 1 reducer. Instead, expose the output dataset via explore.

anew commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansing.java:
```diff
@@ -0,0 +1,70 @@
+      // Properties for Explore (to create a partitioned Hive table)
+      .setEnableExploreOnCreate(true)
+      .setExploreFormat("text")
+      .setExploreFormatProperty("delimiter", "\n")
```
this will use a newline between the key and value. Is that the intention?

anwar6953 replied:
The intent is that the records are newline-separated. I ran an explore query (select *), and the records came out as expected.

anew replied:
have you looked at the output file? I think it may contain bogus records that Hive ignores

anwar6953 replied:
Sorry, what do you mean by this?
I've looked at the output file, and its simply the list of records, one per line.

anew replied:
or maybe empty lines, not sure.

anwar6953 replied:
I didn't see any empty lines in the output file. I think the lines in the output file would correspond to records that are not filtered.

anew replied:
I read the code, and you are right, if the key is null, then neither key nor separator are written at all. So I guess you don't need to configure the separator for the output format.

anwar6953 replied:
Why does this mean I don't need to configure the separator for the output format?

anew replied:
Because the separator is never used. https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/TextOutputFormat.java#L100

anwar6953 replied:
While writing, it's not used, but during explore query, I believe its used as the delimiter to separate the records.
https://github.com/caskdata/cdap/blob/feature/incremental-partition-consumer-dc-example/cdap-explore/src/main/java/co/cask/cdap/explore/service/ExploreTableManager.java#L429-429

anew replied:
I see. I am not quite sure whether it makes sense to have a newline as the field separator. But if that means that the line will never be accidentally split by Hive, ok, let;s keep it here.

anew commented on cdap-examples/DataCleansing/src/main/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduce.java:
```diff
@@ -0,0 +1,152 @@
+  /**
+   * An identity reducer which writes only the values to the file.
+   */
+  public static class IdentityReducer extends Reducer<LongWritable, Text, NullWritable, Text> {
```
do you need a reducer? Why not use a map-only job?

anew replied:
the reducer will group by key, which is the position in the file. If the input is more than one file, there can be more than one line (value) in the reduce input for each key.

anwar6953 replied:
with a map-only task, the records are not the only thing written to the output file.
This IdentityReducer ensures that only the values are written to the file (so as to match the format of the input file).

anew replied:
If your mapper has a null output key, it will behave exactly the same as the reducer. I think it is better not to have a reducer here.

anwar6953 replied:
Ok, I changed the mapper to output null as the key and removed the reducer. Thanks for the heads up.

anew commented on cdap-examples/DataCleansing/src/test/java/co/cask/cdap/examples/datacleansing/DataCleansingMapReduceTest.java:
```diff
@@ -0,0 +1,179 @@
+    MapReduceManager mapReduceManager = applicationManager.getMapReduceManager(DataCleansingMapReduce.NAME).start(args);
+    mapReduceManager.waitForFinish(5, TimeUnit.MINUTES);
+
+    Assert.assertEquals(filterInvalidRecords(RECORDS1), getCleanData(now));
```
Do you validate in ay way that the output looks good? Or only using SQL?

anwar6953 replied:
I filter the records in the test class also, and then compare to the records returned by the explore query.

anew replied:
What I meant is read back the files directly and validate them.

anwar6953 replied:
That's what I was originally doing (serving the file from a user service), but after a discussion with @terence and @sreevatsanraman , we came to the conclusion that we should try not to demonstrate doing that, because it would not scale if the data size is actually large (which a user may try this example with larger data).

anew replied:
I did not say you should have a service. The unit test can instantiate the dataset, get the partition that was written, open the file and read its contents.

anwar6953 replied:
Good point. Done that and pushed.

anew commented:
@anwar6953 please see my responses to your comments

anwar6953 commented:
@anew Pushed updates.

anew commented:
@anwar6953, the map-only MapReduce looks good now. Please see my responses to the other questions.

anew commented:
Also, the output may contain more than one part file. @rohitsinha54 recently had a test case that failed because there were two output parts. Better to iterate over all files beginning with "part".

anwar6953 commented:
Addressed comments, @anew

anew commented:
LGTM
